### PR TITLE
Implemented QIM Mvt Watermark & added unit test project

### PR DESCRIPTION
**Pull request by danlla**

### CODE REVIEW

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+        private int[,] GenerateWinx(int key)
+        {
+            //var r = (int)Math.Floor((double)_options.M * _options.M / sizeMessage);
+            var sizeMessage = (int)Math.Floor((double)_options.M * _options.M / _options.R);
+            var random = new Random(key);
+            var winx = new int[_options.M, _options.M];
+
+            for (int i = 0; i < _options.M; i++)
+                for (int j = 0; j < _options.M; j++)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: var

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+            for (var i = 0; i < _options.Extent; i++)
+                for (var j = 0; j < _options.Extent; j++)
+                    map[i, j] = Convert.ToBoolean(random.Next() % 2);
+            map = ChangeMap(map);
+            return map;
+        }
+
+        private bool[,] ChangeMap(bool[,] map)
+        {
+            var count = 0;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: What is the purpose of this variable?

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+                    return true;
+
+            if (y - 1 >= 0)
+                if (Convert.ToInt32(map[x, y - 1]) != value)
+                    return true;
+
+            return false;
+        }
+
+        private double Statistics(VectorTile tile, Polygon polygon, bool[,] map, Envelope envelopeTile, double extentDist, out int s0, out int s1)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: 1. Nothing will change in this method if you replace Polygon with Geometry. Is it essential to use Polygon?
2. tileEnvelope sounds better than envelopeTile

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+
+            if ((s0 == 0 && s1 == 0) || s0 + s1 < _options.CountPoints)
+                return -1;
+
+            return (double)Math.Abs(s0 - s1) / (s1 + s0);
+        }
+
+        private struct IntPoint
+        {
+            public int x;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: 1. X
2. readonly

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+            public int y;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: 1. Y
2. readonly

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+
+        private List<IntPoint> GetOppositePoint(bool[,] map, int x, int y, int value)
+        {
+            int xRes;
+            int yRes;
+
+            var listPoints = new List<IntPoint>();
+
+            if (x + 1 < _options.Extent)
+                if (Convert.ToInt32(map[x + 1, y]) != value)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: There are a lot of conversions in the code.
Maybe it's better to make "map" variable a matrix of int's? or vice versa make "value" variable a bool?
It's the same in a bunch of other methods

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+        }
+
+        private struct IntPoint
+        {
+            public int x;
+            public int y;
+            public IntPoint(int x, int y) { this.x = x; this.y = y; }
+        }
+
+        private List<IntPoint> GetOppositePoint(bool[,] map, int x, int y, int value)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: IEnumerable<IntPoint>

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+            if (y - 1 >= 0)
+                if (Convert.ToInt32(map[x, y - 1]) != value)
+                {
+                    xRes = x;
+                    yRes = y - 1;
+                    listPoints.Add(new IntPoint(xRes, yRes));
+                }
+            return listPoints;
+        }
+        private List<IntPoint> FindOppositeIndexes(bool[,] map, int value, int x, int y)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Add empty line

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+
+                            if (countSuited % step == 0)
+                            {
+                                var listPoints = FindOppositeIndexes(map, mapValue, x, y);
+
+                                foreach (var point in listPoints)
+                                {
+                                    var geometryCopy = geometry.Copy();
+                                    double xMeteres = envelopeTile.MinX + point.x * extentDist;
+                                    double yMeteres = envelopeTile.MinY + point.y * extentDist;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: 1. var
2. There is a mistake in the word Meteres

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+                                    var coor = CoordinateConverter.MetersToDegrees(new Coordinate(xMeteres, yMeteres));
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Maybe "coord" is better?

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+                                    var tmp = 0;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: This is a terrible name for a variable

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 12]
```diff
@@ -0,0 +1,504 @@
...
+                                            geometryCopy.Coordinates[^1].Y = geometryCopy.Coordinates[0].Y;
+                                        }
+                                        if (!geometryCopy.IsValid)
+                                            continue;
+                                    }
+
+                                    for (var k = j + 1; k < geometryCopy.Coordinates.Length; k++)
+                                    {
+                                        var coordinate = CoordinateConverter.DegreesToMeters(geometryCopy.Coordinates[k]);
+                                        if (coordinate.X == coordinateMeters.X && coordinate.Y == coordinateMeters.Y)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Bad comparison.
Do it like that:
```C#
Math.Abs(coordinate.X - coordinateMeters.X) < tiny const number
```

Example: https://dotnetfiddle.net/Md5p5Q

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 13]
```diff
@@ -0,0 +1,504 @@
...
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j)
+                            }
+                    ));
+
+                    var stat = Statistics(tile, polygon, map, envelopeTile, extentDist, out int s0, out int s1);
+                    if (stat == -1)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Same problem with double comparing as above

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 14]
```diff
@@ -0,0 +1,504 @@
...
+                            {
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j)
+                            }
+                    ));
+
+                    var stat = Statistics(tile, polygon, map, envelopeTile, extentDist, out int s0, out int s1);
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: var in out statement

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 15]
```diff
@@ -0,0 +1,504 @@
...
+                    if (stat == -1)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: See the notes on the previous method

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 16]
```diff
@@ -0,0 +1,504 @@
...
+
+        public VectorTileTree Embed(VectorTileTree tiles, int key, BitArray message)
+        {
+            var countTiles = tiles.Count();
+
+            var countBit = (int)Math.Floor((double)_options.M * _options.M / _options.R);
+            if (countBit > (int)Math.Floor((double)message.Count / countTiles))
+                throw new ArgumentOutOfRangeException();
+
+            int current = 0;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: var

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 17]
```diff
@@ -0,0 +1,504 @@
...
+            _options = options;
+        }
+
+        public VectorTileTree Embed(VectorTileTree tiles, int key, BitArray message)
+        {
+            var countTiles = tiles.Count();
+
+            var countBit = (int)Math.Floor((double)_options.M * _options.M / _options.R);
+            if (countBit > (int)Math.Floor((double)message.Count / countTiles))
+                throw new ArgumentOutOfRangeException();
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: What exactly is the out-of-range method argument here?
Either you specify a specific argument in the constructor parameters, or this type of exception is not very suitable

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs [hunk 18]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs:1]: Cleanup solution

File: MvtWatermark/MvtWatermarkTests/MvtWatermarkTests.csproj [hunk 19]
```diff
@@ -0,0 +1,34 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/MvtWatermarkTests.csproj]: Let's migrate to net6 and enable nullable

File: MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+using System.Collections;
+
+namespace MvtWatermarkTests
+{
+    public class QimMvtWatermarkTests
+    {
+        [Fact]
+        public void OneTileOneBit()
+        {
+            var path = Path.Combine(Directory.GetCurrentDirectory(), "stp.mbtiles");
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs]: How can a third-party developer find this file?

File: MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+            var x = 658;
+            var y = 334;
+            var z = 10;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs]: const int

File: MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+            using var sqliteConnection = new SqliteConnection($"Data Source = {path}");
+            sqliteConnection.Open();
+
+            using var command = new SqliteCommand(@"SELECT tile_data FROM tiles WHERE zoom_level = $z AND tile_column = $x AND tile_row = $y", sqliteConnection);
+            command.Parameters.AddWithValue("$z", z);
+            command.Parameters.AddWithValue("$x", x);
+            command.Parameters.AddWithValue("$y", (1 << z) - y - 1);
+            var bytes = (byte[])command.ExecuteScalar();
+
+            using var memoryStream = new MemoryStream(bytes);
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs]: Check bytes for null

File: MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+            var reader = new MapboxTileReader();
+
+            memoryStream.Seek(0, SeekOrigin.Begin);
+            using var decompressor = new GZipStream(memoryStream, CompressionMode.Decompress, false);
+            var tile = reader.Read(decompressor, new NetTopologySuite.IO.VectorTiles.Tiles.Tile(x, y, z));
+
+            var options = new QimMvtWatermarkOptions(0.6, 0.5, 4096, 20, 2, 10, 100);
+
+            var message = new BitArray(new bool[] {true});
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs]: The word "bool" is not needed here

**Review by kolyanch — CHANGES_REQUESTED** [review]

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs [hunk 24]
```diff
@@ -0,0 +1,24 @@
+﻿namespace MvtWatermark.QimMvtWatermark
+{
+    public class QimMvtWatermarkOptions
```
Comment (vicanfed) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs]: Add Nb
Add joint M, R, Nb initialization by (int?, int?, int?)
Add T1
Make Delta2 private, add Delta1, init them based on k

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 25]
```diff
@@ -0,0 +1,491 @@
...
+
+namespace MvtWatermark.QimMvtWatermark
+{
+    public class QimMvtWatermark : IMvtWatermark
+    {
+        private readonly QimMvtWatermarkOptions _options;
+
+        private int[,] GenerateWinx(int key)
+        {
+            var sizeMessage = (int)Math.Floor((double)_options.M * _options.M / _options.R);
```
Comment (vicanfed) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Calc this value inside the Options class

**Review by vicanfed — CHANGES_REQUESTED** [review]
Ещё нужно:

1. Реализовать второй метод извлечения - с группировкой по `winx`
2. Добавить описание назначений методов и параметров в комментариях в `QimMvtWatermark` и `QimMvtWatermarkOptions`.
3. `key` для VectorTile не должен быть равен `key` для `VectorTileTree`. Он должен генерироваться на основе ID тайла и родительского `key`.

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 26]
```diff
@@ -0,0 +1,583 @@
...
+                return true;
+            if (CheckNearestPoints(map, x, y - i, value))
+                return true;
+
+        }
+        return false;
+    }
+
+    /// <summary>
+    /// Сhecks the nearest points for the opposite value
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: The first letter C in the word Checks is written in Russian

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 27]
```diff
@@ -0,0 +1,583 @@
...
+            {
+                var l = GetOppositePoint(map, x, y + 1, value);
+                listPoints.AddRange(l);
+            }
+
+            if (CheckNearestPoints(map, x, y - i, value))
+            {
+                var l = GetOppositePoint(map, x, y - 1, value);
+                listPoints.AddRange(l);
+            };
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Unnecessary semicolon

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 28]
```diff
@@ -0,0 +1,583 @@
...
+                }
+                feature.Geometry = geometry;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Embeds a message into one vector tile
+    /// </summary>
+    /// <param name="tiles">Vector tile for embedding message</param>
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: *tile

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 29]
```diff
@@ -0,0 +1,583 @@
...
+            }
+        if (!embedded)
+            return null!;
+        return tile;
+    }
+
+    /// <summary>
+    /// Extracts an embedded message from one vector tile
+    /// </summary>
+    /// <param name="tiles">Vector tile to extract the message from</param>
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: *tile

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 30]
```diff
@@ -0,0 +1,583 @@
...
+    }
+
+    /// <summary>
+    /// Embeds a message into one vector tile
+    /// </summary>
+    /// <param name="tiles">Vector tile for embedding message</param>
+    /// <param name="key">Secret key</param>
+    /// <param name="message">The message to embed</param>
+    /// <returns>Vector tile with an embedded message</returns>
+    public VectorTile Embed(VectorTile tile, int key, BitArray message)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: The return type should be nullable VectorTile

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 31]
```diff
@@ -0,0 +1,583 @@
...
+                }
+
+                if (value == 0)
+                {
+                    var countAdded = (int)Math.Ceiling(((s0 + s1) * (_options.T2 + _options.Delta2) + s1 - s0) / 2);
+                    ChangeCoordinate(tile, polygon, envelopeTile, extentDist, map, false, countAdded, s1);
+                }
+            }
+        if (!embedded)
+            return null!;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Null-forgiving operator is not needed if you return nullable

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 32]
```diff
@@ -0,0 +1,583 @@
...
+            {
+                if (dict[i] > 0)
+                    bits[i] = true;
+                if (dict[i] < 0)
+                    bits[i] = false;
+            }
+        }
+
+        if (!embedded)
+            return null!;
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: Null-forgiving operator is not needed if you return nullable

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 33]
```diff
@@ -0,0 +1,583 @@
...
+        return tile;
+    }
+
+    /// <summary>
+    /// Extracts an embedded message from one vector tile
+    /// </summary>
+    /// <param name="tiles">Vector tile to extract the message from</param>
+    /// <param name="key">Secret key</param>
+    /// <returns>Extracted message</returns>
+    public BitArray Extract(VectorTile tile, int key)
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: The return type should be nullable BitArray

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 34]
```diff
@@ -0,0 +1,583 @@
...
+
+                tiles[tileId] = tile;
+                current += _options.Nb;
+            }
+        }
+        return tiles;
+    }
+
+    /// <summary>
+    /// Extracts an embedded message from vectortiletree
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs]: *VectorTileTree

File: MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs [hunk 35]
```diff
@@ -0,0 +1,166 @@
...
+    [Fact]
+    public void VectorTileTreeFromFewTiles()
+    {
+        var path = Path.Combine(Directory.GetCurrentDirectory(), "stp10zoom.mbtiles");
+        const int z = 10;
+
+        using var sqliteConnection = new SqliteConnection($"Data Source = {path}");
+        sqliteConnection.Open();
+
+        var tiletree = new VectorTileTree();
```
Comment (kolyanch) [inline MvtWatermark/MvtWatermarkTests/QimMvtWatermarkTests.cs]: *tileTree

**Review by kolyanch — CHANGES_REQUESTED** [review]

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs [hunk 36]
```diff
@@ -0,0 +1,75 @@
...
+        Delta2 = 0.5 * T2;
+        T1 = 15;
+        Delta1 = (int)Math.Round(0.5 * T1);
+        Extent = 4096;
+        Distance = 2;
+        M = 10;
+        R = 10;
+        Nb = 5;
+        IsGeneralExtractionMethod = false;
+    }
```
Comment (vicanfed) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs]: Just set all the parameters and recall another constructor

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs [hunk 37]
```diff
@@ -0,0 +1,75 @@
...
+    /// <summary>
+    /// The number of embedded bits
+    /// </summary>
+    public int Nb { get; set; }
+    /// <summary>
+    /// Indicates that the extraction will take place over a series of squares
+    /// </summary>
+    public bool IsGeneralExtractionMethod { get; set; }
+
+    public QimMvtWatermarkOptions(double k, double t2, int t1, int extent, int distance,int nb, int r, bool isGeneralExtractionMethod = false)
```
Comment (vicanfed) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermarkOptions.cs]: Rewrite the constructor taking `int? nb, int? r, int? m` where this triplet is valid if the only value is null. In the implementation, check this condition and set the three properties calculating one missing value from two others.

File: MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs [hunk 38]
```diff
@@ -0,0 +1,583 @@
...
+                    if (s0 > s1)
+                        bits[i] = false;
+                }
+            }
+        }
+        else
+        {
+            for (var i = 0; i < _options.Nb; i++)
+            {
+                if (dict[i] > 0)
```
Comment (vicanfed) [inline MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs:494]: Majority voting is not implemented (see the final item in the description of the extraction procedure)

**Review by vicanfed — CHANGES_REQUESTED** [review]

**Pull request merged**